hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/lock/StorageBasedLockProvider.java

```diff
@@ -0,0 +1,581 @@
+  // In reality, within a single cloud provider all nodes share the same ntp
+  // server
+  // therefore we do not expect drift more than a few ms.
+  // However, since our lock leases are pretty long, we can use a high buffer.
+  private static final long CLOCK_DRIFT_BUFFER_MS = 500;
+
+  // When we retry lock upserts, do so 5 times
+  private static final long LOCK_UPSERT_RETRY_COUNT = 5;
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(StorageBasedLockProvider.class);
+
+  Logger logger;
+
+  // The lock service implementation which interacts with storage
+  private final StorageLock lockService;
+
+  private final long heartbeatInterval;
+  private final long lockValidity;
+  private final String ownerId;
+  private final String lockFilePath;
+  private final String bucketName;
+  private final HeartbeatManager heartbeatManager;
+  private transient Thread shutdownThread = null;
+
+  @GuardedBy("this")
+  private StorageLockFile currentLockObj = null;
+  @GuardedBy("this")
+  private boolean isClosed = false;
+
+  private synchronized void setLock(StorageLockFile lockObj) {
+    if (lockObj != null && !Objects.equals(lockObj.getOwner(), this.ownerId)) {
+      throw new HoodieLockException("Owners do not match. Current lock owner: " + this.ownerId + " lock path: "
+          + this.lockFilePath + " owner: " + lockObj.getOwner());
+    }
+    this.currentLockObj = lockObj;
+  }
+
+  /**
+   * Default constructor for StorageBasedLockProvider, required by LockManager
+   * to instantiate it using reflection.
+   * 
+   * @param lockConfiguration The lock configuration, should be transformable into
+   *                          StorageBasedLockConfig
+   * @param conf              Storage config, ignored.
+   */
+  public StorageBasedLockProvider(final LockConfiguration lockConfiguration, final StorageConfiguration<?> conf) {
+    StorageBasedLockConfig config = new StorageBasedLockConfig.Builder()
+        .fromProperties(lockConfiguration.getConfig()).build();
+    heartbeatInterval = config.getHeartbeatPoll();
+    lockValidity = config.getLockValidityTimeout();
+
+    String configuredLocksLocation = config.getLocksLocation();
+
+    // If not configured, recalculate the locks location as .hoodie/.locks;
+    // otherwise (the lock location is configured), the configuration location is used as the folder
+    // to which the lock file is written to, and the lock file name is determined by the table's base path
+    String locksLocation = StringUtils.isNullOrEmpty(configuredLocksLocation)
+        ? String.format("%s%s%s", config.getHudiTableBasePath(), StoragePath.SEPARATOR, LOCKS_FOLDER_NAME)
+        : configuredLocksLocation;
+
+    URI uri = parseURI(locksLocation);
+    bucketName = uri.getHost(); // For most schemes, the bucket/container is the host.
+    String folderName = uri.getPath(); // Path after the bucket/container.
+
+    String fileName = StringUtils.isNullOrEmpty(configuredLocksLocation)
```

Comment:
I thought even if user overrides the config, they will only provide the direction and not the file name. so, why not we dictate the file name internally. 
I mean, irrespective of whether we infer the location of if user overrides, the file name has to be static. just the parent directory can change

Reply:
not using locks location anymore